fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/BulkDataUtils.java

```diff
@@ -145,101 +157,159 @@ private static int getFhirResourceFromBufferReader(BufferedReader resReader, int
+        int parseFailures = 0;
+
+        try {
+            if (transientUserData.getBufferReader() == null) {
+                BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath));
+                transientUserData.setBufferReader(resReader);
+                // Skip the already processed lines after opening the input stream for first read.
+                parseFailures = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), numOfLinesToSkip, fhirResources);
+            } else {
+                parseFailures = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), 0, fhirResources);
             }
+        } catch (Exception ex) {
+            // Clean up.
+            fhirResources.clear();
+            cleanup4TransientUserData(transientUserData, true);
+            // Log the error and throw exception to fail the job, the job can be continued from the current checkpoint after the problem is solved.
+            logger.warning("readFhirResourceFromLocalFile: " + "Error proccesing file " + filePath + " - " + ex.getMessage());
+            throw ex;
         }
-        return exported;
+
+        return parseFailures;
     }
 
 
+    /**
+     * @param dataUrl - URL to the ndjson file.
+     * @param numOfLinesToSkip - number of lines to skip before read.
+     * @param fhirResources - List holds the FHIR resources.
+     * @param transientUserData - transient user data for the chunk.
+     * @return - number of parsing failures.
+     * @throws Exception
+     */
     public static int readFhirResourceFromHttps(String dataUrl, int numOfLinesToSkip, List<Resource> fhirResources,
-            boolean isReuseInput, ImportTransientUserData transientUserData) {
-        int exported = 0;
-        if (isReuseInput) {
+            ImportTransientUserData transientUserData) throws Exception {
+        int parseFailures = 0;
+        int retryTimes = Constants.IMPORT_RETRY_TIMES;
+        do {
             try {
                 if (transientUserData.getBufferReader() == null) {
                     InputStream inputStream = new URL(dataUrl).openConnection().getInputStream();
                     transientUserData.setInputStream(inputStream);
                     BufferedReader resReader = new BufferedReader(new InputStreamReader(inputStream));
                     transientUserData.setBufferReader(resReader);
+                    // Skip the already processed lines after opening the input stream for first read.
+                    parseFailures = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), numOfLinesToSkip, fhirResources);
+                } else {
+                    parseFailures = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), 0, fhirResources);
+                }
+                break;
+            } catch (Exception ex) {
+                // Clean up.
+                fhirResources.clear();
```

Comment:
what does fhirResources.clear accomplish and why?  I see it's passed into this method.

Reply:
wanted to clear before the retry, the retry will reopen the input stream, skip all the processed lines in the last batch, and then redo the whole read for this batch again. let me improve the codes a little bit to skip the success reads in this batch also ... then I will not need to clear it it ...

Reply:
done